### PR TITLE
Modifies `QuickPlot.plot` to take a list of times and show superimposed plots in case of 1D variables.

### LINKED ISSUE #4271: Allow `QuickPlot.plot` to take a list of times
This was part of #3164, but I have encountered a few cases recently where this would have been handy. The suggested API would be for `QuickPlot.plot` to take a list of times and then superimpose the solutions at those times. This would only work for 1D variables. 0D is redundant, the main question is whether we should add multiple vertical lines or not.

### PR DESCRIPTION
**Pull request by medha-14**
# Description
Fixes #4271 
Related to #3164

```
model = pybamm.lithium_ion.DFN()
sim = pybamm.Simulation(model)
sim.solve([0, 3600])
plot = QuickPlot(sim)
plot.plot([0.5,0.6,0.8])
```
image

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by medha-14** [comment]
I will be working on adding different colors and a legend to the time plots to make them easier to differentiate. Please let me know if you have any suggestions for changes or improvements.

**Comment by medha-14** [comment]
I've added different colors to the time plots to make them easier to differentiate. Additionally, I've fixed some of the failing tests. Please suggest any changes or improvements.

image

**Review by arjxn-py — COMMENTED** [review]
Awesome work @medha-14, you might just want to look into the `keyerror` that's being raised. 
You can try debugging that by running one test at a time something like - `pytest tests/unit/test_plotting/test_quick_plot.py::test_plot_1plus1D_spme
` Otherwise this looks really good.

**Review by arjxn-py — COMMENTED** [review]
Awesome work @medha-14, you might just want to look into the `keyerror` that's being raised. 
You can try debugging that by running one test at a time something like - `pytest tests/unit/test_plotting/test_quick_plot.py::test_plot_1plus1D_spme
` Otherwise this looks really good.

File: src/pybamm/plotting/quick_plot.py
Comment (kratman) [inline src/pybamm/plotting/quick_plot.py:541]:
```suggestion
                            color=self.colors[i],
```

File: tests/unit/test_plotting/test_quick_plot.py
Comment (kratman) [inline tests/unit/test_plotting/test_quick_plot.py:127]: See above, but I don't see why we shouldn't be able to just pass in zero here

**Review by kratman — CHANGES_REQUESTED** [review]

**Comment by medha-14** [comment]
I will work on fixing the failing tests. Should I also start working on adding new tests for this method?

**Comment by medha-14** [comment]
I have fixed the failing tests and tried to add tests ,  using a list of time in the plot function. I am open to work on any suggested improvements.

**Comment by medha-14** [comment]
Hi @kratman, could you review this and let me know what further changes I should make?

**Comment by medha-14** [comment]
Are there any other changes or improvements that I could make to further enhance the implementation?

**Comment by kratman** [comment]
@medha-14 Sorry for the delay on this one, I will start reviewing this again

File: src/pybamm/plotting/quick_plot.py
Comment (kratman) [inline src/pybamm/plotting/quick_plot.py:497]: Probably easier to do it this way
```suggestion
        if not isinstance(t, list):
```

File: src/pybamm/plotting/quick_plot.py
Comment (kratman) [inline src/pybamm/plotting/quick_plot.py:487]: Can you also add the type annotation to the function signature?

File: src/pybamm/plotting/quick_plot.py
Comment (kratman) [inline src/pybamm/plotting/quick_plot.py:600]: The index does not appear to be used, so can we just simplify this?
```suggestion
                for t_single in t:
```

**Review by kratman — COMMENTED** [review]
Small findings, I will do a full review another day

**Comment by kratman** [comment]
Please also add a change log entry. Note that we are doing a release, so once #4598 is merged, there will likely be a merge conflict in the changelog

File: CHANGELOG.md
Comment (kratman) [inline CHANGELOG.md:23]: This will have to move to the unreleased section due to the release this morning

File: src/pybamm/plotting/quick_plot.py
Comment (kratman) [inline src/pybamm/plotting/quick_plot.py:320]: Guessing this was a typo
```suggestion
                        f"'{variable_tuple[0]}' has domain '{domain}', but '{variable_tuple[idx]}' has domain '{variable.domain}'"
```

File: src/pybamm/plotting/quick_plot.py
Comment (kratman) [inline src/pybamm/plotting/quick_plot.py:488]: This needs to be indented to render properly
```suggestion
            Dimensional time (in 'time_units') at which to plot. Can be a single time or a list of times.
```

File: src/pybamm/plotting/quick_plot.py
Comment (kratman) [inline src/pybamm/plotting/quick_plot.py:482]: The python 3.9 failure comes from this line, but your change is correct.

Add
```
from __future__ import annotations
```
To the very top of the file

**Review by kratman — CHANGES_REQUESTED** [review]
This should be the last round of findings. Once these are resolved I will approve. Please double check that the figures still appear as you originally intended. Just good to do a visual double check. I will review those as well

File: src/pybamm/plotting/quick_plot.py
Comment (kratman) [reply src/pybamm/plotting/quick_plot.py:488]: You can see this if you check the "readthedocs" output in the checks section. It is good to check the docs whenever you update them

File: CHANGELOG.md
Comment (agriyakhetarpal) [inline CHANGELOG.md:5]:
```suggestion
- Modified `quick_plot.plot` to accept a list of times and generate superimposed graphs for specified time points. ([#4529](https://github.com/pybamm-team/PyBaMM/pull/4529))
```

(nit)

**Review by agriyakhetarpal — DISMISSED** [review]
Thanks, @medha-14! I only skimmed through the changes and haven't been involved in the review here, I'll request a review from Eric for the final merge.

**Review by kratman — APPROVED** [review]
Thanks looks good

**Comment by kratman** [comment]
@allcontributors add @medha-14 for code

**Comment by kratman** [comment]
I will merge once I see the windows 3.9 check pass. I do not expect a problem, so this should be merged in a couple minutes

**Comment by medha-14** [comment]
Thanks a lot everyone!

**Pull request merged**
